### PR TITLE
SCEP Support

### LINKED ISSUE #5896: packetfence-pki: configure Organisation, Country, State, etc. fields on Templates
**Is your feature request related to a problem? Please describe.**
When you create a certificate template, you don't have possibility to set Organisation, Country, State, etc. fields.

Consequently, you have to set this field each time you generate a certificate based on template.

**Describe the solution you'd like**
Inherit Organisation, Country, State, etc. fields from template when creating certificate with possibility to override with certificate values. 

**Describe alternatives you've considered**
Configure Organisation, Country, State, etc. fields on Templates in place of Certificates.

### LINKED ISSUE #5897: packetfence-pki: add Organisation unit field in template and certificate
**Is your feature request related to a problem? Please describe.**
Organizational unit (OU) is a field allowed in X509 certificate according to this [post](https://stackoverflow.com/questions/6464129/certificate-subject-x-509#6464435) but we don't have possibility to configure it at the moment.

It's an interesting field if you want to extract it from certificate to match against an organizational unit on your AD for example. 

**Describe the solution you'd like**
Allow to configure Organizational unit (OU) for generated certificates, using templates (see #5896).

**Describe alternatives you've considered**
Allow to configure other fields according to [RFC 5280](https://tools.ietf.org/html/rfc5280) and updates.

### LINKED ISSUE #6021: SAN attribute in the PacketFence PKI
It would be nice to support SAN in the PF PKI in order to sign HTTPS connection.

### LINKED ISSUE #6022: Support Server Name Indication (SNI) on captive portal
**Is your feature request related to a problem? Please describe.**
We support different FQDN on captive portal, for example in a tenant setup. But we don't support having several SSL certificates for each FQDN on captive portal. It means you have to use a wildcard certificate and all your tenants should use same domain name (which is, most of the time, not the case).

**Describe the solution you'd like**
Support Server Name Indication (SNI) in haproxy-portal.

### PR DESCRIPTION
**Pull request by fdurand**
# Description
Added support for SCEP.
Added dns/ip in certificate
Added template in certificate profile

# Impacts
pfpki

# Issue
related to #6022
fixes #6021
fixes #5897
fixes #5896

# Delete branch after merge
YES

# Checklist
- [ ] Document the feature
- [ ] Add unit tests
- [ ] Add acceptance tests (TestLink)

# NEWS file entries
## New Features
* pfpki scep support

## Enhancements
* DNS/IP attributes in certificate (pfpki)
* Additional template attributes in certificate profile

## Bug Fixes
* fixes #6022
* fixes #6021
* fixes #5897
* fixes #5896

### CODE REVIEW

File: db/upgrade-X.X.X-X.Y.Z.sql [hunk 1]
```diff
@@ -163,6 +163,39 @@ CREATE TABLE IF NOT EXISTS `remote_clients` (
...
+    ADD COLUMN IF NOT EXISTS `scep_challenge_password` varchar(255) AFTER p12_mail_footer,
+    ADD COLUMN IF NOT EXISTS `scep_allow_renewal` varchar(255) AFTER scep_challenge_password;
+
+
+\! echo "altering pki_cas"
+ALTER TABLE pki_cas
+    ADD COLUMN IF NOT EXISTS `organisational_unit` varchar(255) DEFAULT NULL AFTER organisation,
+    ADD COLUMN IF NOT EXISTS `ocsp_url` varchar(255) DEFAULT NULL AFTER issuer_name_hash;
+
+\! echo "altering pki_cert"
```
Comment (nqb) [inline db/upgrade-X.X.X-X.Y.Z.sql]:
```suggestion
\! echo "altering pki_certs"
```

File: go/caddy/pfpki/certutils/certutils.go [hunk 2]
```diff
@@ -267,3 +271,87 @@ func parseKeyFile(filename string) (interface{}, error) {
...
+
+	b, err := x509.DecryptPEMBlock(pemBlock, password)
+	if err != nil {
+		return x509.ParsePKCS1PrivateKey(pemBlock.Bytes)
+	}
+
+	return x509.ParsePKCS1PrivateKey(b)
+}
+
+// load an encrypted private key from disk
```
Comment (nqb) [inline go/caddy/pfpki/certutils/certutils.go]: Wrong comment

File: go/caddy/pfpki/pfpki.go [hunk 3]
```diff
@@ -86,53 +78,70 @@ func buildPfpkiHandler(ctx context.Context) (Handler, error) {
...
+
+	api.Handle("/pki/scep/{id}", handlers.ManageSCEP(PFPki)).Methods("GET", "POST")
+
+	api.Handle("/scep/", handlers.ManageSCEP(PFPki)).Methods("GET", "POST")
+
+	api.Handle("/scep/{id}/", handlers.ManageSCEP(PFPki)).Methods("GET", "POST")
+
+	api.Handle("/pki/scep/", handlers.ManageSCEP(PFPki)).Methods("GET", "POST")
+
+	api.Handle("/pki/scep/{id}/", handlers.ManageSCEP(PFPki)).Methods("GET", "POST")
```
Comment (nqb) [inline go/caddy/pfpki/pfpki.go]: It looks like content is present twice.

File: go/caddy/pfpki/types/types.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+		get() error
+		revoke() error
+	}
+
+	// Error struct
+	Error struct {
+		Field   string `json:"field"`
+		Message string `json:"message"`
+	}
+	// Error struct
```
Comment (nqb) [inline go/caddy/pfpki/types/types.go]:
```suggestion
	// Errors struct
```

File: html/pfappserver/root/src/views/Configuration/pki/profiles/_components/TheForm.vue [hunk 5]
```diff
@@ -40,6 +65,10 @@
           :column-label="$i18n.t('Extended key usage')"
           :text="$i18n.t('Optional. One or many of: serverAuth, clientAuth, codeSigning, emailProtection, timeStamping, msCodeInd, msCodeCom, msCTLSign, msSGC, msEFS, nsSGC.')"
         />
+        <form-group-ocsp-url namespace="ocsp_url"
+          :column-label="$i18n.t('OCSP Url')"
+          :text="$i18n.t('Optional. This is the url of the OCSP server that will be added in the certificate.If empty then the ca one will be used')"
```
Comment (nqb) [inline html/pfappserver/root/src/views/Configuration/pki/profiles/_components/TheForm.vue]:
```suggestion
          :text="$i18n.t('Optional. This is the url of the OCSP server that will be added in the certificate. If empty then the ca one will be used')"
```

**Review by nqb — CHANGES_REQUESTED** [review]
### Coding style

I see a change in naming of functions (i.e. `generatePassword` becomes `GeneratePassword`). 
Is it something we will generate ?

**Comment by fdurand** [comment]
> ### Coding style
> I see a change in naming of functions (i.e. `generatePassword` becomes `GeneratePassword`).
> Is it something we will generate ?

It's because the function needs to be exported, it's why some of them has been changed.

File: html/pfappserver/root/src/views/Configuration/pki/profiles/_components/TheForm.vue [hunk 6]
```diff
@@ -40,6 +65,10 @@
           :column-label="$i18n.t('Extended key usage')"
           :text="$i18n.t('Optional. One or many of: serverAuth, clientAuth, codeSigning, emailProtection, timeStamping, msCodeInd, msCodeCom, msCTLSign, msSGC, msEFS, nsSGC.')"
         />
+        <form-group-ocsp-url namespace="ocsp_url"
+          :column-label="$i18n.t('OCSP Url')"
```
Comment (nqb) [inline html/pfappserver/root/src/views/Configuration/pki/profiles/_components/TheForm.vue]:
```suggestion
          :column-label="$i18n.t('OCSP URL')"
```

File: html/pfappserver/root/src/views/Configuration/pki/profiles/_components/TheForm.vue [hunk 7]
```diff
@@ -40,6 +65,10 @@
...
+          :text="$i18n.t('Optional. This is the url of the OCSP server that will be added in the certificate. If empty then the ca one will be used')"
```
Comment (nqb) [inline html/pfappserver/root/src/views/Configuration/pki/profiles/_components/TheForm.vue]:
```suggestion
          :text="$i18n.t('Optional. This is the URL of the OCSP server that will be added in the certificate. If empty then the ca one will be used')"
```

**Review by nqb — CHANGES_REQUESTED** [review]

**Review by nqb — CHANGES_REQUESTED** [review]
### [FIXED] Inheritance of fields from template

Certs contains fields inherited from template but this inheritance doesn't appear in GUI.
It looks like we don't receive these informations from API when clicking on certificate.

**Comment by fdurand** [comment]
> ### Inheritance of fields from template
> Certs contains fields inherited from template but this inheritance doesn't appear in GUI.
> It looks like we don't receive these informations from API when clicking on certificate.

It's in the certificate but not in the DB, let me fix that.

**Review by nqb — CHANGES_REQUESTED** [review]
### scep_allow_renewal

I'm confused with this field.
Based on name, I will expect a boolean value to enable/disable this feature but in GUI, field description is: 
> Number of days before SCEP authorize renewal

**Comment by fdurand** [comment]
> ### scep_allow_renewal
> I'm confused with this field.
> Based on name, I will expect a boolean value to enable/disable this feature but in GUI, field description is:
> 
> > Number of days before SCEP authorize renewal

In fact it represent the period before the certificate expire to allow the renewal.

**Comment by nqb** [comment]
> In fact it represent the period before the certificate expire to allow the renewal.

What do you think about a rename to `scep_days_before_renewal` with following description: "Number of days before SCEP authorize renewal" ?

**Comment by fdurand** [comment]
> > In fact it represent the period before the certificate expire to allow the renewal.
> 
> What do you think about a rename to `scep_days_before_renewal` with following description: "Number of days before SCEP authorize renewal" ?

yep let me change that.

**Comment by nqb** [comment]
As discussed with @fdurand, I got some issues to get certificate using `sscep`. @fdurand is looking at that.

**Comment by fdurand** [comment]
> As discussed with @fdurand, I got some issues to get certificate using `sscep`. @fdurand is looking at that.

Just did a test with fresh package:

```
[root@facketfence pf]# sscep getca  -u http://192.168.2.1/scep/intune -c ./ca-prefix -i Inverse -v -d
sscep: starting sscep, version 0.6.1
sscep: new transaction
sscep: transaction id: SSCEP transactionId
sscep: hostname: 192.168.2.1
sscep: directory: scep/intune
sscep: port: 80
sscep: SCEP_OPERATION_GETCA
sscep: requesting CA certificate
sscep: scep msg: GET /scep/intune?operation=GetCACert&message=Inverse HTTP/1.0

sscep: server returned status code 200
sscep: MIME header: application/x-x509-ca-cert
sscep: valid response from server
sscep: MD5 fingerprint: 66:F4:7D:53:31:15:14:7C:E7:B2:7C:8A:C8:6F:7C:B8
sscep: CA certificate written as ./ca-prefix
```

So it looks to work.
As we talked, it looks that it depend of devices , i did it with sscep, iphone, android and all 2 worked.

The only difference is that you are running debian and me centos.
I will try on debian.

**Comment by fdurand** [comment]
> > As discussed with @fdurand, I got some issues to get certificate using `sscep`. @fdurand is looking at that.
> 
> Just did a test with fresh package:
> 
> ```
> [root@facketfence pf]# sscep getca  -u http://192.168.2.1/scep/intune -c ./ca-prefix -i Inverse -v -d
> sscep: starting sscep, version 0.6.1
> sscep: new transaction
> sscep: transaction id: SSCEP transactionId
> sscep: hostname: 192.168.2.1
> sscep: directory: scep/intune
> sscep: port: 80
> sscep: SCEP_OPERATION_GETCA
> sscep: requesting CA certificate
> sscep: scep msg: GET /scep/intune?operation=GetCACert&message=Inverse HTTP/1.0
> 
> sscep: server returned status code 200
> sscep: MIME header: application/x-x509-ca-cert
> sscep: valid response from server
> sscep: MD5 fingerprint: 66:F4:7D:53:31:15:14:7C:E7:B2:7C:8A:C8:6F:7C:B8
> sscep: CA certificate written as ./ca-prefix
> ```
> 
> So it looks to work.
> As we talked, it looks that it depend of devices , i did it with sscep, iphone, android and all 2 worked.
> 
> The only difference is that you are running debian and me centos.
> I will try on debian.

I am able to replicate on Debian.

**Comment by fdurand** [comment]
> > > As discussed with @fdurand, I got some issues to get certificate using `sscep`. @fdurand is looking at that.
> > 
> > 
> > Just did a test with fresh package:
> > ```
> > [root@facketfence pf]# sscep getca  -u http://192.168.2.1/scep/intune -c ./ca-prefix -i Inverse -v -d
> > sscep: starting sscep, version 0.6.1
> > sscep: new transaction
> > sscep: transaction id: SSCEP transactionId
> > sscep: hostname: 192.168.2.1
> > sscep: directory: scep/intune
> > sscep: port: 80
> > sscep: SCEP_OPERATION_GETCA
> > sscep: requesting CA certificate
> > sscep: scep msg: GET /scep/intune?operation=GetCACert&message=Inverse HTTP/1.0
> > 
> > sscep: server returned status code 200
> > sscep: MIME header: application/x-x509-ca-cert
> > sscep: valid response from server
> > sscep: MD5 fingerprint: 66:F4:7D:53:31:15:14:7C:E7:B2:7C:8A:C8:6F:7C:B8
> > sscep: CA certificate written as ./ca-prefix
> > ```
> > 
> > 
> > So it looks to work.
> > As we talked, it looks that it depend of devices , i did it with sscep, iphone, android and all 2 worked.
> > The only difference is that you are running debian and me centos.
> > I will try on debian.
> 
> I am able to replicate on Debian.

Tried with the latest code of sscep ( https://github.com/certnanny/sscep ) and it works like a charm.
Let me try to package it and include as a dependencie in this PR.

**Comment by nqb** [comment]
Using new `sscep` package on Debian, I'm able to get a certificate based on a CSR generated locally on my PacketFence server running Debian. My CSR contains a SCEP challenge password which was defined in my PKI profile.

Resulting certificate contains key usage and extended key usage defined in my PKI profile but OCSP URL is missing.

**Comment by nqb** [comment]
> Resulting certificate contains key usage and extended key usage defined in my PKI profile but OCSP URL is missing.

As discussed with @fdurand, OCSP URL should be defined in CSR.
It seems that EKU and KU are taken from PKI profile but it looks like they could be defined in CSR too.
SCEP is only used to sign.

**Pull request merged**